src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt

```diff
@@ -45,54 +44,76 @@ private val RsTraitItem.typeParamSingle: TyTypeParameter?
     get() =
         typeParameterList?.typeParameterList?.singleOrNull()?.let { TyTypeParameter.named(it) }
 
-class ImplLookup(private val project: Project, private val items: StdKnownItems) {
+const val DEFAULT_RECURSION_LIMIT = 64
+
+class ImplLookup(
+    private val project: Project,
+    private val items: StdKnownItems,
+    private var _ctx: RsInferenceContext? = null
+) {
     // Non-concurrent HashMap and lazy(NONE) are safe here because this class isn't shared between threads
     private val primitiveTyHardcodedImplsCache = mutableMapOf<TyPrimitive, Collection<BoundElement<RsTraitItem>>>()
+    private val binOpsTraitAndOutputCache = mutableMapOf<ArithmeticOp, Pair<RsTraitItem, RsTypeAlias>?>()
     private val fnTraits by lazy(NONE) {
         listOf("fn", "fn_mut", "fn_once").mapNotNull { RsLangItemIndex.findLangItem(project, it) }
     }
+    private val fnOnceOutput: RsTypeAlias? by lazy(NONE) {
+        val trait = RsLangItemIndex.findLangItem(project, "fn_once") ?: return@lazy null
+        findAssociatedType(trait, "Output")
+    }
     val fnOutputParam by lazy(NONE) {
-        RsLangItemIndex.findLangItem(project, "fn_once")?.let { findFreshAssociatedType(it, "Output") }
+        fnOnceOutput?.let { TyTypeParameter.associated(it) }
     }
     private val copyTrait: RsTraitItem? by lazy(NONE) {
         RsNamedElementIndex.findDerivableTraits(project, "Copy").firstOrNull()
     }
-    private val derefTraitAndTarget: Pair<RsTraitItem, TyTypeParameter>? = run {
+    private val derefTraitAndTarget: Pair<RsTraitItem, RsTypeAlias>? = run {
         val trait = RsLangItemIndex.findLangItem(project, "deref") ?: return@run null
-        findFreshAssociatedType(trait, "Target")?.let { trait to it }
+        findAssociatedType(trait, "Target")?.let { trait to it }
+    }
+    private val indexTraitAndOutput: Pair<RsTraitItem, RsTypeAlias>? by lazy(NONE) {
+        val trait = RsLangItemIndex.findLangItem(project, "index") ?: return@lazy null
+        findAssociatedType(trait, "Output")?.let { trait to it }
+    }
+    private val iteratorTraitAndOutput: Pair<RsTraitItem, RsTypeAlias>? by lazy(NONE) {
+        val trait = items.findIteratorTrait() ?: return@lazy null
+        findAssociatedType(trait, "Item")?.let { trait to it }
     }
+    private val intoIteratorTraitAndOutput: Pair<RsTraitItem, RsTypeAlias>? by lazy(NONE) {
+        val trait = items.findCoreItem("iter::IntoIterator") as? RsTraitItem ?: return@lazy null
+        findAssociatedType(trait, "Item")?.let { trait to it }
+    }
+
+    private val ctx: RsInferenceContext
+        get() {
+            if (_ctx == null)
+                _ctx = RsInferenceContext()
```

Comment:
Braces or `ctrl+shilf+j` or `get() = _ctx ?: RsInferenceContext().apply { _ctx = this }`